crates/mpc-protocols/src/mpc_service.rs

```diff
@@ -0,0 +1,115 @@
+}
+
+impl MPCService {
+    pub fn new() -> Self {
+        Self {
+            mpc_instances: HashMap::new(),
+            pending: VecDeque::new(),
+            active_instances_counter: 0,
+        }
+    }
+
+    /// Spawns an asynchronous task to handle incoming messages for a new MPC instance.
+    /// The [`MPCService`] will forward any message related to that instance to this channel.
+    fn spawn_mpc_messages_handler(&self, mut receiver: mpsc::Receiver<MPCInput>) {
+        tokio::spawn(async move {
+            while let Some(message) = receiver.recv().await {
+                // TODO (#235): Implement MPC messages handling
+            }
+        });
+    }
+
+    /// Filter the relevant MPC events from the transaction events & handle them
+    pub fn handle_mpc_events(&mut self, events: &Vec<Event>) -> anyhow::Result<()> {
+        for event in events {
+            if CreatedProofMPCEvent::type_() == event.type_ {
+                let deserialized_event: CreatedProofMPCEvent = bcs::from_bytes(&event.contents)?;
+                self.handle_proof_init_event(deserialized_event);
+                println!("event: CreatedProofMPCEvent {:?}", event);
+            };
+        }
+        Ok(())
+    }
+
+    /// Handles a proof initialization event
+    /// Spawns a new MPC instance if the number of active instances is below the limit
+    /// Otherwise, adds the instance to the pending queue
+    fn handle_proof_init_event(&mut self, event: CreatedProofMPCEvent) {
+        println!(
+            "Received start flow event for session ID {:?}",
+            event.session_id
+        );
+        // let active_instances_counter = self.active_instances_counter.clone().into_inner();
+        // If the number of active instances exceeds the limit, add to pending
+        if self.active_instances_counter >= MAX_ACTIVE_MPC_INSTANCES {
+            self.mpc_instances.insert(
+                event.session_id.clone().bytes,
+                MPCInstance {
+                    status: MPCStatus::Pending,
+                    messages_handler_sender: None,
+                    pending_messages: vec![],
+                },
+            );
+            self.pending.push_back(event.session_id.bytes);
+            return;
+        }
+        let (messages_handler_sender, messages_handler_receiver) = mpsc::channel(100);
+        self.spawn_mpc_messages_handler(messages_handler_receiver);
+        let _ = messages_handler_sender.send(MPCInput::InitEvent(event.clone()));
+        self.mpc_instances.insert(
+            event.session_id.clone().bytes,
+            MPCInstance {
+                status: MPCStatus::Active,
+                messages_handler_sender: Some(messages_handler_sender),
+                pending_messages: vec![],
+            },
+        );
+        self.active_instances_counter += 1;
+        println!(
+            "Added MPCInstance to service for session_id {:?}",
+            event.session_id
+        );
```

Comment:
1. Consider making this an `if-else` instead.
2. Consider moving the `self.mpc_instances.insert(` outside the `if-else` to avoid unnecessary duplication.
3. How unique are these session_ids? This function implicitly assumes that the `session_id.bytes` are not yet in the map.

Reply:
1. Zeev repeatedly asks us to return early instead of using `if-else`.
2. They have different fields. 
3. `session_id` is move `UID` type so it is unique.

Reply:
I don't want to go against Zeev's wishes, but I think we can resolve this there being a need for returning early.
What do you think of the following code?

```rs
/// Handles a proof initialization event
/// Spawns a new MPC instance if the number of active instances is below the limit
/// Otherwise, adds the instance to the pending queue
fn handle_proof_init_event(&mut self, event: CreatedProofMPCEvent) {
    info!(
        "Received start flow event for session ID {:?}",
        event.session_id
    );

    let mut new_instance = MPCInstance {
        status: MPCSessionStatus::Pending,
        input_receiver: None,
        pending_messages: vec![],
    };
    self.mpc_instances
        .insert(event.session_id.clone().bytes, new_instance);

    // Activate the instance if possible
    if self.active_instances_counter < MAX_ACTIVE_MPC_INSTANCES {
        new_instance.status.set_active();
        new_instance.init_input_receiver();
        self.active_instances_counter += 1;        
    } else {
        self.pending_instances_queue
            .push_back(event.session_id.bytes);
    };

    info!(
        "Added MPCInstance to service for session_id {:?}",
        event.session_id
    );
}
```